include/rnp/rnp2.h

```diff
@@ -0,0 +1,283 @@
+
+/* TODO: should these be NULL-terminated hex strings, or just
+ the raw data? Document which. */
+rnp_result_t rnp_key_get_fprint(rnp_key_t key, char **fprint);
+rnp_result_t rnp_key_get_keyid(rnp_key_t key, char **keyid);
+rnp_result_t rnp_key_get_grip(rnp_key_t key, char **grip);
+
+rnp_result_t rnp_key_is_locked(rnp_key_t key, bool *result);
+rnp_result_t rnp_key_unlock(rnp_key_t key, rnp_passphrase_cb cb, void *app_ctx);
+
+rnp_result_t rnp_key_is_protected(rnp_key_t key, bool *result);
+rnp_result_t rnp_key_protect(rnp_key_t key, rnp_passphrase_cb cb, void *app_ctx);
+rnp_result_t rnp_key_unprotect(rnp_key_t key, rnp_passphrase_cb cb, void *app_ctx);
+
+rnp_result_t rnp_key_is_primary_key(rnp_key_t key, bool *result);
+rnp_result_t rnp_key_is_subkey(rnp_key_t key, bool *result);
+rnp_result_t rnp_key_is_secret(rnp_key_t key, bool *result);
+rnp_result_t rnp_key_is_public(rnp_key_t key, bool *result);
+
+/* TODO: function to add a userid to a key */
+
+/* Signature/verification operations */
+
+/**
+* Generate an embedded signature (the output will include the
+* message contents)
+* @param keyring the keyring
+* @param ident the key to sign with
+* @param hash_fn the hash function to use
+* @param msg the message to sign
+* @param msg_len the length of msg in bytes
+* @param sig on success, the output signature buffer
+* @param sig_len on success, the length of sig in bytes
+*/
+rnp_result_t rnp_sign(rnp_keyring_t keyring,
+                      const char *  ident,
+                      const char *  hash_fn,
+                      bool          clearsign,
+                      bool          armor,
+                      const uint8_t msg[],
+                      size_t        msg_len,
+                      uint8_t **    sig,
+                      size_t *      sig_len);
+
+/**
+* Generate an embedded signature (the output will include the
+* message contents)
+* @param keyring the keyring
+* @param ident the key to sign with
+* @param hash_fn the hash function to use
+* @param msg the message to sign
+* @param msg_len the length of msg in bytes
+* @param sig on success, the output signature buffer
+* @param sig_len on success, the length of sig in bytes
+*/
+rnp_result_t rnp_sign_detached(rnp_keyring_t keyring,
+                               const char *  ident,
+                               const char *  hash_fn,
+                               bool          armor,
+                               const uint8_t msg[],
+                               size_t        msg_len,
+                               uint8_t **    sig,
+                               size_t *      sig_len);
+
+/**
+* Verify a signature with embedded message
+* @param key the keyring
+* @param sig the signature
+* @param sig_len length of signature in bytes
+* @param msg on succes, output buffer to the message
+* @param msg_len on success, length of msg in bytes
+*/
+
+rnp_result_t rnp_verify(
```

Comment:
I believe we should add some 'keyneeded' callback instead of keyring here and in decryp() function. So we could search for the key (on PGP servers, multiple files, database, etc) instead of loading the whole known bunch of keys into memory.

Reply:
I think the key store may eventually evolve to provide this functionality per the discussion in https://github.com/riboseinc/rnp/issues/351#issuecomment-320001558., though we're a long way from that currently.

I think that adding a "storage provider/backend" concept to the key store might be a potential way of handling it, for example. Then a "find key" operation on a key store could query the database backend, etc. (Probably better discussed in the linked issue)

@ronaldtse For an immediate solution, would a 'keyneeded' callback (for verify and decrypt operations) be necessary for what you need?

Reply:
Actually looking closer at some of these interfaces I see other similar issues and it may be easiest to get this callback in place. Let me see what we can do.

Reply:
@dewyatt yes the callback would work already. The "storage provider" / "backend" concept is exactly what we need.